### PR TITLE
Needs Design Review: First pass at fixing the formatting of VB Line Continuation Formatting

### PR DESCRIPTION
**Pull request by paul1956**
This addresses _ on line without code, with or without comment. This uses existing Formatting Rules which don't understand VB Line Continuation on blank lines or followed by comment. The results need a design review. I temporarily renamed the effected tests by added 3 _, they could be removed but it would be harder to see the effects of any changes, these are the only tests that I found that hit the code I changed. One area that needs review is the alignment of lines under lines with _ comment.

### CODE REVIEW

File: src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb [hunk 1]
```diff
@@ -812,16 +812,16 @@ End Class</Code>
             Dim expected = <Code>Class C
     Sub Method(Optional ByVal i As Integer = 1)
         Dim aa = 1 + _
-                 2 + _
-                 3
+                     2 + _
+                         3
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: this formatting doesn't seem desirable.

Comment (paul1956) [reply src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: It is not desirable and I would love to fix it but I don't see a rule that would format it correctly and doesn't undo all the newer formatting. I have experimented with the available rules and you either get this behavior or the "2" is left aligned sometimes and correctly aligned others depending on what the line looks like before the formatting.

Comment (paul1956) [reply src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: The code that does the indenting based on the rules needs to understand " _ ' Optional Comment' and I have not found it yet or I need to use a different rule.

**Comment by CyrusNajmabadi** [comment]
> they could be removed but it would be harder to see the effects of any changes, 

Please don't do this.  It appears as if your code changes are regressing these scenarios.  Removing tests will simply hide that and allow regressions in areas we have tests explicitly there to ensure the behavior of.

> these are the only tests that I found that hit the code I changed. One area that needs review is the alignment of lines under lines with _ comment.

I think you need to update your change so that it can give you the behavior you want without regressing these cases.

**Comment by CyrusNajmabadi** [comment]
I don't see any tests actually showing cases you think should be fixed/improved by your product code change.  Can you write tests that at least show what positive behavior you think the changes so far have?

File: src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb [hunk 2]
```diff
@@ -69,7 +69,15 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Formatting
 
             ' line continuation
             If trivia2.Kind = SyntaxKind.LineContinuationTrivia Then
-                Return LineColumnRule.ForceSpacesOrUseAbsoluteIndentation(spacesOrIndentation:=1)
+                If trivia1.Kind = SyntaxKind.None Then
+                    Return LineColumnRule.ForceSpacesOrUseAbsoluteIndentation(spacesOrIndentation:=1)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb]: what code example causes this case to be hit?

File: src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb [hunk 3]
```diff
@@ -69,7 +69,15 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Formatting
...
+                ElseIf trivia1.Kind = SyntaxKind.LineContinuationTrivia Then
+                    Return LineColumnRule.ForceSpacesOrUseDefaultIndentation(spaces:=0)
+                ElseIf trivia1.Kind = SyntaxKind.commenttrivia Then
+                    Return LineColumnRule.ForceSpacesOrUseDefaultIndentation(spaces:=-1)
+                Else
+                    Return LineColumnRule.PreserveLinesWithFollowingPrecedingIndentation()
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb]: can you provide comments in this code giving examples of what each of these if-blocks would correspond to and why the LineColumnRules you are returning are desirable?  For example, i have no idea why you are returning `spacesOrIndentation:=1` for `trivia1.Kind = SyntaxKind.None` or why you return `spaces:=-1` for `trivia1.Kind = SyntaxKind.commenttrivia`

File: src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: > I would love to fix it but I don't see a rule that would format it correctly and doesn't undo all the newer formatting

What do you mean by 'newer formatting'?

> The code that does the indenting based on the rules needs to understand " _ ' Optional Comment' and I have not found it yet or I need to use a different rule.

Can you write a set of tests that show the behavior you want roslyn to have, but doesn't have?  That way we can then have a code change that makes those tests pass, while not regressing these other scenarios?

File: src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb [hunk 4]
```diff
@@ -833,21 +833,21 @@ End Class</Code>
...
 
         <Fact, Trait(Traits.Feature, Traits.Features.Formatting)>
-        Public Async Function LineContinuation4() As Task
+        Public Async Function LineContinuation4___() As Task
             Dim code = <Code>Class C
     Sub Method(Optional ByVal i As Integer = 1)
                     Dim aa = 1 + _
             _
-                                          _
+                                          _' Comment 1
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: please don't change the inputs of existing tests.  Instead, add a new test with the different input.  By doing this, i can't see what the impact of your code change is on hte original input of the original test.

File: src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb [hunk 5]
```diff
@@ -3600,7 +3624,7 @@ End Class
 
         <Fact, Trait(Traits.Feature, Traits.Features.Formatting)>
         <WorkItem(529899, "DevDiv_Projects/Roslyn")>
-        Public Async Function IndentContinuedLineOfSingleLineLambdaToFunctionKeyword() As Task
+        Public Async Function IndentContinuedLineOfSingleLineLambdaToFunctionKeyword___() As Task
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: why did you add ___ to all the tests?

File: src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb (hunk 4 above)
Comment (paul1956) [reply src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: Yes I realized I needed to do that. Some of the old tests will fail but I think for now that is OK.

File: src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb (hunk 5 above)
Comment (paul1956) [reply src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: So I could identify all the tests that are effected by my changes, as I said it was temporary. I will put the old tests back (which was always my intent) and put my changes next to them with ___.

File: src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb (hunk 2 above)
Comment (paul1956) [reply src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb]: A file with only _

File: src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb (hunk 3 above)
Comment (paul1956) [reply src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb]: Done is newest commit

File: src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb (hunk 5 above)
Comment (paul1956) [reply src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: I have restored all the only test, there are 3 possibilities, New  ___ tests passes and old test fails this is expected, the new test is what I think the code should look like.  Both tests fail, here the old test is expected to fail but my changes did not produce desired result as shown in new test. Old test fails and there is no new test, my changes broke the old code.

File: src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb (hunk 1 above)
Comment (paul1956) [reply src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: That is my goal, I restored all the original tests, some fail is expected other fail because my code incorrectly breaks them. Is there a way to expect the old test to fail and have it pass?

**Comment by paul1956** [comment]
The main thing I am trying to change is what happens when a _ with optional Comment is on a line by itself and what happens on the code line following after that line(s). Today lines after, align left with 1 space. This is similer to how in VB aligns Directives Vs. C#. All of my attempts to date have broken the code followed by _ ' Comment on a single line case or caused the newly indented lines to drift right by two space when a comment is present.

File: src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb]: please document this code to make those cases clear.

File: src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/VisualBasicTest/Formatting/FormattingTests.vb]: thanks!

File: src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb [hunk 6]
```diff
@@ -69,7 +69,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Formatting
...
+                If trivia1.Kind = SyntaxKind.None Then
+                    ' File or line or trivia starting with _ requires a leading space
+                    Return LineColumnRule.ForceSpacesOrUseAbsoluteIndentation(spacesOrIndentation:=1)
+                ElseIf trivia1.Kind = SyntaxKind.LineContinuationTrivia Then
+                    ' _ line followed by another _ and the _ should align
+                    Return LineColumnRule.ForceSpacesOrUseDefaultIndentation(spaces:=0)
+                ElseIf trivia1.Kind = SyntaxKind.CommentTrivia Then
+                    ' _ ' Comment aligns with the ' instead of the _ causing each line to move right
+                    ' 2 space
+                    Return LineColumnRule.ForceSpacesOrUseDefaultIndentation(spaces:=-1)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Formatting/Engine/Trivia/VisualBasicTriviaFormatter.vb]: so, in terms of a comment, this isn't really helpful.  What i'd like is both an example of waht the original VB code looks like that this case is trying to handle, what we want the final code to look like, and why this particular method/arguent is the right thing to do that.  i.e. why do we use `ForceSpacesOrUseDefaultIndentation`?  why do we pass `spaces:=-1` to it?  

I honestly don't know the answers, and reviewing the PR isn't informative enough.  As such, the code needs effective comments to make it clear.

**Comment by CyrusNajmabadi** [comment]
Tagging @heejaechang .  @heejaechang can you help @paul1956 here?

**Comment by paul1956** [comment]
Ping @sharwell, I started looking at this again and if I can't get some help I should close. It seems that since I started working on this new code fixes and tests have been added that depend on the current behavior of _ and they break with my PR and it is unclear to me if I need to fix their tests (becuase my new formatting is better) or my code (because the current formatting is better/desired) and as has been said the behavior of my code is sometimes better but sometimes odd or undesirable.

**Comment by sharwell** [comment]
I added several tests for the current state of line continuation support in #44267. I'm not sure we need to implement handling beyond this, but I at least wanted to capture the test cases. 👍
